crates/circuit/src/circuit_instruction.rs

```diff
@@ -432,6 +432,41 @@ impl CircuitInstruction {
         matrix.map(|mat| mat.into_pyarray_bound(py).into())
     }
 
+    fn to_matrix(&self, py: Python) -> Option<PyObject> {
+        self.matrix(py)
+    }
+
+    fn __array__(
+        &self,
+        py: Python,
+        dtype: Option<PyObject>,
+        copy: Option<bool>,
+    ) -> PyResult<PyObject> {
+        if copy == Some(false) {
+            return Err(PyValueError::new_err(
+                "A copy is needed to return an array from this object.",
+            ));
+        }
+        let res =
+            match self.to_matrix(py) {
+                Some(res) => res,
+                None => return Err(PyTypeError::new_err(
+                    "ParameterExpression with unbound parameters cannot be represented as an array",
```

Comment:
Is this the only way `self.to_matrix(py)` can fail - can we forward the error message produced by python here?

Reply:
Yeah, the `matrix()` method defined in the `Operation` trait only returns an `Option<Array2<Complex64>>` . The goal of the trait was to define pure rust based semantics for querying operations. So for standard gates it's either a matrix if all the parameters are floats or `None` if any parameter is a `ParameterExpression`. For the `PyGate` rust space wrapper it converting any potential error (or the lack of a matrix) to `None`: https://github.com/Qiskit/qiskit/blob/main/crates/circuit/src/operations.rs#L1549-L1565.

We could add an alternative path that tried to preserve the error message from Python, but the point of the `Operation` trait was to abstract some of this away for a rust context. Realistically though I modeled this error message after what happens in the common case:

```
>>> from qiskit.circuit import Parameter
>>> from qiskit.circuit.library import RZGate
>>> gate = RZGate(Parameter('a'))
>>> gate.__array__()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/tmp/test_11/lib/python3.12/site-packages/qiskit/circuit/library/standard_gates/rz.py", line 139, in __array__
    ilam2 = 0.5j * float(self.params[0])
                   ^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/test_11/lib/python3.12/site-packages/qiskit/circuit/parameterexpression.py", line 415, in __float__
    raise TypeError(
TypeError: ParameterExpression with unbound parameters (dict_keys([Parameter(a)])) cannot be cast to a float.
```

There is a loss of information for opaque gates that don't define a matrix where if a user did
```python
from qiskit.circuit import Gate, CircuitInstruction
custom_gate = Gate(name='foo', num_qubits=2, params=[])
custom_gate.__array__()
custom_instruction = CircuitInstruction(custom_gate)
custom_instruction.__array__()
```
The `custom_instruction.__array__()` path would return this parameter expression error vs an `AttributeError` for the custom_gate.__array__()` call.

That all being said this path doesn't really get called in this PR, as custom gates will work in the python domain. We only use this `to_matrix()` and `__array__` if it's a standard gate: https://github.com/Qiskit/qiskit/pull/12701/files/fa774b3e3d493d9b04cb98166e40673301abae10#diff-f652dda983f1a9918c329ea373b9e570cc2f037ab22c59a968424deea2610832R70
but I guess it's a potential future hazard for ourselves and maybe we want to avoid doing this. The other alternative I guess would be to type check in `commute_matmul()` and call `DAGOpNode.matrix` to build the `Operator` instead of passing the `DAGOpNode` directly to `Operator`. I went back and forth on this when I wrote the PR. Which would you prefer?

Reply:
I think the current version is fine, thanks for explaining! :-)

Reply:
Actually after writing that all out, I've convinced myself that adding `to_matrix()` and `__array__` is the incorrect path forward here. It is too big an interface to add haphazardly and it's easy enough to workaround in the commutation checker code directly. I'll refactor this as part of the rebase I need to do to fix merge conflicts anyway.

Reply:
Removed in: https://github.com/Qiskit/qiskit/pull/12701/commits/5f94bf07ee0d38a40e02f0befec048a204c44737